### PR TITLE
fork returns a Process.  Small resource leak fixed.

### PR DESCRIPTION
**Pull request by technorama**

### CODE REVIEW

File: src/process/waitpid_future.cr [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    @value = nil
+  end
+
+  def value
+    receive
+
+    if value = @value
+      return value
+    else
+      raise "waitpid channel closed"
```
Comment (jhass) [inline src/process/waitpid_future.cr]: Can we find a more specific exception here?

File: src/process/process.cr [hunk 2]
```diff
@@ -88,14 +112,9 @@ class Process
...
   def self.waitpid pid
-    chan = Event::SignalChildHandler.instance.waitpid pid
-    status = chan.receive
-    if status
-      return status
-    else
-      raise "waitpid channel closed"
-    end
+    {{ raise "DEPRECATED: fork and Process.new return a Process.  Use process.wait instead." }}
   end
```
Comment (jhass) [inline src/process/process.cr]: I think if we drop this and don't let `Process.kill` accept pids anymore, we should also drop `Process.kill` and make the primary interface `Process#kill` instead.

**Comment by technorama** [comment]
Process.kill(pid) is needed. You may want to kill another process not created by fork.
```
pid = File.read("process.pid").to_i
Process.kill sig, pid
```

Waitpid is different.  You can only wait on forked processes.  Passing an arbitrary pid would only be necessary if a) you use LibC.fork (which shouldn't be supported) or b) an external library forks.

I can remove the Process.pid(Process) if you like.

File: src/process/run.cr [hunk 3]
```diff
@@ -137,11 +137,18 @@ class Process
...
+
     fork_input.try &.close
     fork_output.try &.close
     fork_error.try &.close
   end
 
+  def initialize @pid
+    @waitpid_future = WaitpidFuture.new @pid
+    @wait_count = 0
+  end
```
Comment (jhass) [inline src/process/run.cr]: So this should be internal, that is private or protected?

**Comment by jhass** [comment]
I don't see `Process.pid(Process)` defined anywhere?

**Comment by technorama** [comment]
I meant `Process.kill(Process)`.

File: src/process/waitpid_future.cr (hunk 1 above)
Comment (technorama) [inline src/process/waitpid_future.cr]: I'm not sure what that should be.  It only occurs when in the child after forking while another fiber is waiting for different process to exit (before forking).

I could return nil but then the type changes to an optional.  
Or raise a custom exception?

Suggestions?

Comment (jhass) [inline src/process/waitpid_future.cr]: `Channel::Closed` perhaps?

**Comment by waj** [comment]
I really like `fork` returning Process! But I wouldn't remove `Process.kill(pid)`. I'd remove `Process.kill(Process)` instead and leave `Process#kill`

Don't leave deprecation notices. At this stage we want to be able to go back and forth with any API while we polish them until 1.0.

Do we really need the future object? If it's just used internally it's just another object instance created to execute a process.

File: src/process/process.cr [hunk 4]
```diff
@@ -48,15 +48,39 @@ class Process
       ret = LibC.kill(pid.to_i32, signal.value)
       raise Errno.new(ret) if ret < 0
     end
-    0
+    nil
+  end
+
+  def self.kill(signal : Signal, *procs : Process)
```
Comment (waj) [inline src/process/process.cr]: Do we really need to receive an enumeration of processes here? As said before, I'd remove this method and leave `Process#kill`.

**Comment by technorama** [comment]
The future object is to encapsulate the limited functionality.  The process module is getting rather large.

It could be replaced with
```
  delay { chan.receive }
```

Perhaps it's time to discuss more advanced concurrency primitives like futures.

**Pull request merged**